tests/lib/external/snapd-testing-tools/tools/repack-kernel

```diff
@@ -0,0 +1,231 @@
+    unsquashfs -d "$target/kernel" "$snap_file"
+
+    kver="$(get_kver "$target/kernel")"
+
+    # repack initrd magic, beware
+    # assumptions: initrd is compressed with LZ4, cpio block size 512, microcode
+    # at the beginning of initrd image
+
+    # XXX: ideally we should unpack the initrd, replace snap-boostrap and
+    # repack it using ubuntu-core-initramfs --skeleton=<unpacked> this does not
+    # work and the rebuilt kernel.efi panics unable to start init, but we
+    # still need the unpacked initrd to get the right kernel modules
+    objcopy -j .initrd -O binary "$target"/kernel/kernel.efi "$target/work/initrd"
+
+    # copy out the kernel image for create-efi command
+    objcopy -j .linux -O binary "$target"/kernel/kernel.efi "$target/work/vmlinuz-$kver"
+
+    cp -a "$target"/kernel/kernel.efi "$target/backup/"
+
+    # this works on 20.04 but not on 18.04
+    unmkinitramfs "$target"/work/initrd "$target"/work/unpacked-initrd
+    
+    # copy the unpacked initrd to use as the target skeleton
+    cp -ar "$target/work/unpacked-initrd" "$target/skeleton"
+
+    echo "prepared workspace at $target"
+    echo "  kernel:              $target/kernel ($kver)"
+    echo "  kernel.efi backup:   $target/backup/kernel.efi"
+    echo "  temporary artifacts: $target/work"
+    echo "  initramfs skeleton:  $target/skeleton"
+
+}
+
+prepare() {
+    local target="$1"
+
+    if [ -z "$target" ]; then
+        echo "repack-kernel: missing target for prepare"
+        exit 1
+    fi
+
+    target=$(realpath "$target")
+    local kver="$(get_kver "$target/kernel")"
+
+    (
+        # all the skeleton edits go to a local copy of distro directory
+        local skeletondir="$target/skeleton"
+
+        cd "$target/work"
+        # XXX: need to be careful to build an initrd using the right kernel
+        # modules from the unpacked initrd, rather than the host which may be
+        # running a different kernel
+        (
+            # accommodate assumptions about tree layout, use the unpacked initrd
+            # to pick up the right modules
+            cd unpacked-initrd/main
+            ubuntu-core-initramfs create-initrd \
+                                  --kernelver "$kver" \
+                                  --skeleton "$skeletondir" \
+                                  --feature main \
+                                  --kerneldir "lib/modules/$kver" \
+                                  --output "$target/work/repacked-initrd"
+        )
+
+        # assumes all files are named <name>-$kver
+        ubuntu-core-initramfs create-efi \
+                              --kernelver "$kver" \
+                              --initrd repacked-initrd \
+                              --kernel vmlinuz \
+                              --output repacked-kernel.efi
+
+        cp "repacked-kernel.efi-$kver" "$target/kernel/kernel.efi"
+
+        # XXX: needed?
+        chmod +x "$target/kernel/kernel.efi"
```

Comment:
I can confirm it is not. At least for qemu :)

Reply:
Thanks Alfonso!